src/storage/src/hummock/state_store.rs

```diff
@@ -307,18 +308,50 @@ impl HummockStorage {
             if level.table_infos.is_empty() {
                 continue;
             }
-            let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key));
-            for table_info in table_infos {
-                let table = self
-                    .sstable_store
-                    .sstable(table_info.id, &mut stats)
-                    .await?;
-                table_counts += 1;
-                if let Some(v) = self
-                    .get_from_table(table, &internal_key, key, check_bloom_filter, &mut stats)
-                    .await?
-                {
-                    return Ok(v);
+            match level.level_type() {
+                LevelType::Overlapping | LevelType::Unspecified => {
+                    let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key));
+                    for table_info in table_infos {
+                        let table = self
+                            .sstable_store
+                            .sstable(table_info.id, &mut stats)
+                            .await?;
+                        table_counts += 1;
+                        if let Some(v) = self
+                            .get_from_table(
+                                table,
+                                &internal_key,
+                                key,
+                                check_bloom_filter,
+                                &mut stats,
+                            )
+                            .await?
+                        {
+                            return Ok(v);
+                        }
+                    }
+                }
+                LevelType::Nonoverlapping => {
+                    let mut table_info_idx = level.table_infos.partition_point(|table| {
+                        let ord =
+                            user_key(&table.key_range.as_ref().unwrap().left).cmp(key.as_ref());
+                        ord == Ordering::Less || ord == Ordering::Equal
+                    });
+                    if table_info_idx == 0 {
+                        continue;
+                    }
+                    table_info_idx = table_info_idx.saturating_sub(1);
+                    let table = self
```

Comment:
After we get `table_info_idx`, I think we should further check whether the key falls in the key range of SST in case that the SST key range is not continuous and the key falls in the gap between the key range of two SSTs.

Reply:
Yes. In the worse case, we just run a useless query on sstable.

Reply:
> After we get `table_info_idx`, I think we should further check whether the key falls in the key range of SST in case that the SST key range is not continuous and the key falls in the gap between the key range of two SSTs.
ok, I've added the check for it.